### PR TITLE
client: always enable TCP keepalives with OS defaults

### PR DESCRIPTION
**Pull request by easwars**
In https://github.com/grpc/grpc-go/pull/6672, we attempted to configure OS defaults for TCP keepalive parameters (keepalive time and interval). But this had the unintended consequence of disabling TCP keepalives completely because of the way things are implemented in the Go stdlib. In this PR, we attempt to do the following:

Summary of changes:
- Always enable TCP keepalives on client connections 
  - This has been the behavior for a while because the Go stdlib enables TCP keepalives as long as the `net.Dialer.KeepAlive` field is not set to a negative value. 
  - Explicitly setting the socket option ensures the old behavior persists.
- Use OS defaults for TCP keepalive interval and time
  - We do this by setting the `net.Dialer.KeepAlive` field to a negative value, thereby disabling the Go stdlib's override of these values to `15s`.
 
There will be a follow-up PR for the server side.

Addresses https://github.com/grpc/grpc-go/issues/6250

RELEASE NOTES:
- client: always enable TCP keepalives with OS defaults

### CODE REVIEW

**Comment by easwars** [comment]
@atollena : If you are interested in this change, please let me know and I'll mark you as a reviewer. Thanks.

File: internal/tcp_keepalive.go [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+		// the values of TCP keepalive time and interval. It also prevents the
+		// Go stdlib from enabling TCP keepalives by default.
+		KeepAlive: time.Duration(-1),
+		// This method is called after the underlying network socket is created,
+		// but before dialing the socket (or calling its connect() method). The
+		// combination of unconditionally enabling TCP keepalives here, and
+		// disabling the overriding of TCP keepalive parameters by setting the
+		// KeepAlive field to a negative value above, results in OS defaults for
+		// the TCP keealive interval and time parameters.
+		ControlContext: func(_ context.Context, _, _ string, c syscall.RawConn) error {
```
Comment (atollena) [inline internal/tcp_keepalive.go:45]: feel free to ignore: you can use `Control` instead of `ControlContext` for one less unused parameter.

**Review by atollena — APPROVED** [review]
That's what I had in mind, thanks!

File: dialoptions.go [hunk 2]
```diff
@@ -401,10 +401,12 @@ func WithTimeout(d time.Duration) DialOption {
 // returned by f, gRPC checks the error's Temporary() method to decide if it
 // should try to reconnect to the network address.
 //
-// Note: As of Go 1.21, the standard library overrides the OS defaults for
-// TCP keepalive time and interval to 15s.
-// To retain OS defaults, use a net.Dialer with the KeepAlive field set to a
-// negative value.
+// Note: As of Go 1.21, the standard library overrides the OS defaults for TCP
```
Comment (dfawley) [inline dialoptions.go:404]: I agree this wording is confusing, and have been trying to think of something better...  Maybe:

`Note: All supported releases of Go (as of December 2023) override the OS defaults for...`

Or "as of the Go 1.21 release" or "up to and including at least Go 1.21" for the parenthetical?

File: server.go [hunk 3]
```diff
@@ -813,15 +813,6 @@ func (l *listenSocket) Close() error {
 // Serve returns when lis.Accept fails with fatal errors.  lis will be closed when
 // this method returns.
 // Serve will return a non-nil error unless Stop or GracefulStop is called.
-//
-// Note: As of Go 1.21, the standard library overrides the OS defaults for
```
Comment (dfawley) [inline server.go:817]: Why is this deleted?

File: dialoptions.go [hunk 4]
```diff
@@ -401,10 +401,12 @@ func WithTimeout(d time.Duration) DialOption {
...
+// keepalive time and interval to 15s. To enable TCP keepalive with OS defaults
+// for keepalive time and interval, use a net.Dialer that sets the KeepAlive
+// field set to a negative value, and sets the SO_KEEPALIVE socket option to
```
Comment (dfawley) [inline dialoptions.go:407]: "sets (the KeepAlive field) set to a negative".  Remove the 2nd "set".

File: internal/tcp_keepalive.go (hunk 1 above)
Comment (dfawley) [reply internal/tcp_keepalive.go:45]: Sounds like a minor improvement to me, but I don't feel strongly.

Comment (easwars) [reply internal/tcp_keepalive.go:45]: I went with `ControlContext` over `Control` (even though the context parameter is unused) since the former is preferred *if* both are set.

File: server.go (hunk 3 above)
Comment (easwars) [reply server.go:817]: Because I will be sending a follow-up PR to fix things on the server-side.

File: dialoptions.go (hunk 4 above)
Comment (easwars) [reply dialoptions.go:407]: Done.

File: dialoptions.go (hunk 2 above)
Comment (easwars) [reply dialoptions.go:404]: Done. Thanks.

File: internal/tcp_keepalive.go (hunk 1 above)
Comment (easwars) [reply internal/tcp_keepalive.go]: Looks like the `ControlContext` field was only added in Go1.20, and currently the least supported Go version is still Go1.19. So I'm changing this back to `Control`.

File: server.go (hunk 3 above)
Comment (easwars) [reply server.go:817]: Updated this with the latest commit, based on our discussion on the issue.

File: internal/transport/http2_client.go [hunk 5]
```diff
@@ -264,7 +263,7 @@ func newHTTP2Client(connectCtx, ctx context.Context, addr resolver.Address, opts
 	}
 	keepaliveEnabled := false
 	if kp.Time != infinity {
-		if err = syscall.SetTCPUserTimeout(conn, kp.Timeout); err != nil {
+		if err = isyscall.SetTCPUserTimeout(conn, kp.Timeout); err != nil {
```
[the remark this answers is not in the document]
Comment (easwars) [reply internal/transport/http2_client.go:266]: TCP_USER_TIMEOUT specifies the maximum amount of time that transmitted data may remain unacknowledged before TCP will forcibly close the corresponding connection and return ETIMEDOUT to the application. If the option value is specified as 0, TCP will to use the system default. 

I think it is totally a reasonable thing to use it without TCP keepalives.

As per [A18](https://github.com/grpc/proposal/blob/master/A18-tcp-user-timeout.md#background), the reason we set TCP_USER_TIMEOUT when gRPC keepalives are configured is to ensure that if things are stuck at the TCP layer, setting the TCP_USER_TIMEOUT will ensure that the connection is closed anyways (even if gRPC keepalives are not able to do the same).

Are you concerned about the case where a user specifies a custom `net.Dialer` to disable TCP keepalives, *and* configures gRPC keepalives, and therefore, we end up setting TCP_USER_TIMEOUT, but TCP keepalives are disabled? Even in this case, I feel it should be fine, since TCP_USER_TIMEOUT can work irrespective of whether TCP keepalives are configured on the connection.

[the remark this answers is not in the document]
Comment (dfawley) [reply internal/transport/http2_client.go:266]: Ah no, sorry that makes sense, I'm not sure what I was thinking yesterday.. :)

**Review by dfawley — APPROVED** [review]

**Comment by andyliuliming** [comment]
looks like this PR blocks bazel referencing the grpc module.
